src/jdk.internal.le/share/classes/jdk/internal/org/jline/terminal/impl/AbstractPty.java

```diff
@@ -102,4 +153,103 @@ private void setNonBlocking() {
         }
     }
 
+    private static FileDescriptorCreator fileDescriptorCreator;
+
+    protected static FileDescriptor newDescriptor(int fd) {
+        if (fileDescriptorCreator == null) {
+            String str =
+                    System.getProperty(PROP_FILE_DESCRIPTOR_CREATION_MODE, PROP_FILE_DESCRIPTOR_CREATION_MODE_DEFAULT);
+            String[] modes = str.split(",");
+            IllegalStateException ise = new IllegalStateException("Unable to create FileDescriptor");
+            for (String mode : modes) {
+                try {
+                    switch (mode) {
+                        case PROP_FILE_DESCRIPTOR_CREATION_MODE_NATIVE:
+                            fileDescriptorCreator = null;//new NativeFileDescriptorCreator();
+                            break;
+                        case PROP_FILE_DESCRIPTOR_CREATION_MODE_REFLECTION:
+                            fileDescriptorCreator = new ReflectionFileDescriptorCreator();
+                            break;
+                    }
+                } catch (Throwable t) {
+                    // ignore
+                    ise.addSuppressed(t);
+                }
+                if (fileDescriptorCreator != null) {
+                    break;
+                }
+            }
+            if (fileDescriptorCreator == null) {
+                throw ise;
+            }
+        }
+        return fileDescriptorCreator.newDescriptor(fd);
+    }
+
+    interface FileDescriptorCreator {
+        FileDescriptor newDescriptor(int fd);
+    }
+
+    /*
+     * Class that could be used on OpenJDK 17.  However, it requires the following JVM option
+     *   --add-exports java.base/jdk.internal.access=ALL-UNNAMED
+     * so the benefit does not seem important enough to warrant the problems caused
+     * by access the jdk.internal.access package at compile time, which itself requires
+     * custom compiler options and a different maven module, or at least a different compile
+     * phase with a JDK 17 compiler.
+     * So, just keep the ReflectionFileDescriptorCreator for now.
+     *
+    static class Jdk17FileDescriptorCreator implements FileDescriptorCreator {
+        private final jdk.internal.access.JavaIOFileDescriptorAccess fdAccess;
+        Jdk17FileDescriptorCreator() {
+            fdAccess = jdk.internal.access.SharedSecrets.getJavaIOFileDescriptorAccess();
+        }
+
+        @Override
+        public FileDescriptor newDescriptor(int fd) {
+            FileDescriptor descriptor = new FileDescriptor();
+            fdAccess.set(descriptor, fd);
+            return descriptor;
+        }
+    }
```

Comment:
Since this is a part of the JDK, `jdk.internal.access.JavaIOFileDescriptorAccess` can be used instead of reflection.

Reply:
As this originates in the JLine itself, I would rather avoid doing changes that are unnecessary. Because as a consequence, we would need to re-apply the changes for every upgrade. I believe this code is practically not used in our context, so I don't think it really matters to make it nice.